johnnyfive/gmail.py

```diff
@@ -0,0 +1,404 @@
+        if not self.label_list:
+            # Get the list of labels for the "me" account (API: users.labels.list)
+            results = self.service.users().labels().list(userId='me').execute()
+            self.label_list = results.get('labels', [])
+
+        # If there are no labels, return None
+        if not self.label_list:
+            print('Whoops, no labels found.')
+            return None
+
+        label_id = None
+        # Go through the labels, and return the ID matching the name
+        for label in self.label_list:
+            if label['name'] == name:
+                label_id = label['id']
+
+        return label_id
+
+
+# Newer OAUTH Routines =======================================================#
+def setup_gmail():
+    """setup_gmail Initialize the GMail API (via OAuth)
+
+    [extended_summary]
+
+    NOTE: The first time this is run on a machine, it will open a webpage for
+          authorizing the API.  All subsequent runs will be silent.
+
+    Returns
+    -------
+    `googleapiclient.discovery.Resource`
+        The GMail API service object for consumption by other routines
+    """
+    creds = None
+    if os.path.exists(
+            token_fn := utils.Paths.config.joinpath('gmail_token.json') ):
+        creds = Credentials.from_authorized_user_file(token_fn, SCOPES)
+
+    # If there are no (valid) credentials available, lauch browser to log in
+    if not creds or not creds.valid:
+        if creds and creds.expired and creds.refresh_token:
+            creds.refresh(Request())
+        else:
+            flow = InstalledAppFlow.from_client_secrets_file(
+                utils.Paths.config.joinpath('gmail_credentials.json'), SCOPES)
+            creds = flow.run_local_server(port=0)
+        # Save the credentials for the next run
+        with open(token_fn, 'w', encoding='utf-8') as token:
+            token.write(creds.to_json())
+
+    # Try logging into the GMail API.  If error, print error & return None
+    try:
+        # Call the Gmail API
+        return build('gmail', 'v1', credentials=creds)
+    except HttpError as error:
+        # TODO(developer) - Handle errors from gmail API.
+        print(f'An error occurred: {error}')
+        return None
+
+
+# Utility Functions ==========================================================#
+def build_query(after_date=None, before_date=None):
+    """build_query Build the query string for users.messages.list
+
+    _extended_summary_
+
+    Parameters
+    ----------
+    after_date : `str`
+        Date after which to search for messages.
+    before_date : `str`
+        Date before which to search for messages.
+
+    Returns
+    -------
+    `str`
+        The appropriate query string
+    """
+    q = ''
+    if after_date:
+        q = q + f" after:{after_date}"
+    if before_date:
+        q = q + f" before:{before_date}"
+    return q
+
+
+# Testing Driver =============================================================#
+def main():
```

Comment:
Should pull this sort of stuff out before merging, and put it into an example folder in the root of the repo or something

Reply:
Done.